src/plonk/lookup/prover.rs

```diff
@@ -0,0 +1,547 @@
+            .domain
+            .coeff_to_extended(permuted_input_poly.clone(), Rotation(-1));
+
+        let permuted_input_blind = Blind(C::Scalar::random());
+        let permuted_input_commitment = params
+            .commit_lagrange(&permuted_input_value, permuted_input_blind)
+            .to_affine();
+
+        let permuted_table_poly = pk.vk.domain.lagrange_to_coeff(permuted_table_value.clone());
+        let permuted_table_coset = pk
+            .vk
+            .domain
+            .coeff_to_extended(permuted_table_poly.clone(), Rotation::default());
+        let permuted_table_blind = Blind(C::Scalar::random());
+        let permuted_table_commitment = params
+            .commit_lagrange(&permuted_table_value, permuted_table_blind)
+            .to_affine();
+
+        let permuted = Permuted {
+            permuted_input_value,
+            permuted_input_poly,
+            permuted_input_coset,
+            permuted_input_inv_coset,
+            permuted_input_blind,
+            permuted_input_commitment,
+            permuted_table_value,
+            permuted_table_poly,
+            permuted_table_coset,
+            permuted_table_blind,
+            permuted_table_commitment,
+        };
+
+        self.permuted = Some(permuted.clone());
+        permuted
+    }
+
+    fn permute_column_pair(
+        input_value: &Polynomial<C::Scalar, LagrangeCoeff>,
+        table_value: &Polynomial<C::Scalar, LagrangeCoeff>,
+    ) -> (
+        Polynomial<C::Scalar, LagrangeCoeff>,
+        Polynomial<C::Scalar, LagrangeCoeff>,
+    ) {
+        let mut input_coeffs = input_value.get_values().to_vec();
+        let table_coeffs = table_value.get_values().to_vec();
+
+        // Sort input lookup column values
+        input_coeffs.sort();
+        let permuted_input_value = Polynomial::new(input_coeffs.to_vec());
+
+        // A BTreeMap of each unique element in the table column and its count
+        let mut leftover_table_map: BTreeMap<C::Scalar, u32> =
+            table_coeffs.iter().fold(BTreeMap::new(), |mut acc, coeff| {
+                *acc.entry(*coeff).or_insert(0) += 1;
+                acc
+            });
+        let mut repeated_input_rows = vec![];
+        let mut permuted_table_coeffs = vec![C::Scalar::zero(); table_coeffs.len()];
+
+        for row in 0..input_coeffs.len() {
+            let input_value = input_coeffs[row];
+
+            // If this is the first occurence of `input_value` in the input column
+            if row == 0 || input_value != input_coeffs[row - 1] {
+                permuted_table_coeffs[row] = input_value;
+                // Remove one instance of input_value from leftover_table_map
+                if let Some(count) = leftover_table_map.get_mut(&input_value) {
+                    assert!(*count > 0);
+                    *count -= 1;
+                } else {
+                    // Panic if input_value not found
+                    panic!("Input value not found in table.");
```

Comment:
I think we should return a `ConstraintSystemFailure` error here rather than panicking. I know that means refactoring a bunch of things to return `Result`s, sorry. But it's more consistent with how we handle constraint errors elsewhere, and it's an error made by the caller rather than an unreachable condition in this implementation (unlike the assertions and the unwrap).

Reply:
Hmm, `ConstraintSystemFailure` [only seems to be used in the verifier](https://github.com/zcash/halo2/search?q=ConstraintSystemFailure). That's not what I expected based on its documentation. In any case, I think there should be an error for the situation where a constraint violation is detected during proving.